### ISSUE TITLE
Improve performance for large files

### ISSUE THREAD

**Issue by mtlynch** [mtlynch/picoshare#355]
Reddit user /u/bdpna [reports](https://www.reddit.com/r/selfhosted/comments/tirbdq/picoshare_a_minimalist_easytohost_service_for/iz5ecdc/) that PicoShare is taking 10 hours to upload a 7 GB file to a well-resourced server (Supermicro 4u, dual XEON CPU, 64 GB RAM). PicoShare doesn't seem to be pegging the CPU or RAM.

They tried [increasing the buffer size](https://github.com/mtlynch/picoshare/commit/0704ef17d3f895379ea3087d2c981f0f3f1af1cf) of SQLite data entries, but that had negligible or negative impact on performance.

Other ideas:

* Run Go profiling tools during a large file upload (similar to what I did [for memory exhaustion](https://mtlynch.io/notes/picoshare-perf/))
* Try tuning SQLite via [different pragmas](https://phiresky.github.io/blog/2020/sqlite-performance-tuning/) ([full list](https://www.sqlite.org/pragma.html))

**Comment by joost00719** [comment]
I am also having performance issues. It's stuck on "File uploaded! Processing..." for a very long time when mounted on an NFS share.
The initial upload goes really quick (No clue where it temporary stores it), and then it is processing which then sends it to my NFS share which is on the same physical machine. Speed is about 12-20MByte/s.

I assume the file is written to a temp directory first, and when it's fully uploaded it'll start to process which is just a copy to the sqlite database. The /data directory is on an NFS which causes the slowdown for me I assume.

I'd rather see a direct upload to the /data folder (maybe have a /data/tmp folder) and not have it stored in the sqlite database (just a path reference would be plenty I assume). For this a cleanup routine should be considered (maybe if it's not in the db, delete it from disk, and if it's not on disk, handle it on the front end (e.g. file not found on disk)

**Comment by mtlynch** [comment]
Thanks for reporting this, @joost00719!

Unfortunately, SQLite isn't going to work well if it's over an NFS mount, so I think even if performance were better, you risk data corruption.

At this point, I'm unlikely to move the data out of SQLite. Keeping all data + metadata in SQLite means that Litestream handles all replication for us. If we stored the data as files, we'd have to manage our own replication logic.

I understand that this is an unusual design choice, but that's the design that works best for my use-case, and I optimize PicoShare for scenarios I use.

**Comment by marshalleq** [comment]
I'm having a similar issue, uploading a 15GB file, it uploads fine, But then just disappears with a red empty banner at the bottom.  Kind of defeats the purpose if you cant share a large file.

**Comment by bumgarb** [comment]
I'm not sure if this will be helpful to this issue.  
I'm running PicoShare on a Synology DS1520+, 8GB RAM, 
I have 1G symmetric internet that was testing over 900MB every hour during this upload.
The uploader has a 300MB symmetric but I do not know the conditions on their end at the time of upload.

I had someone upload a 194.87GB file to a Guest Link.  They said they did not get any "valid" confirmation but instead got what looked like an error - a bunch of red text down the page in what looked like HTML to them.  They did not send a screenshot.  

I could tell something was processing on my end as I could see the database slowly increasing in size.  I stopped watching it around 5pm.  I could see the file listed in PicoShare > Files this morning. 

However, I cannot download it from PicoShare directly or through the generated links that I can see in my PicoShare admin console.  I also got a red text error when attempting to delete the file which was basically that deletion failed.  However, after about an hour, I noticed the file no longer appeared in PicoShare > Files.  

I'm not sure how to better handle such a large file successfully.
Also, the PicoShare database is still over 194GB in size.  
Will a prune or purge happen at some point to decrease the size of my database now that the file no longer appears in the list.?Thanks!

**Comment by mtlynch** [comment]
@JitteryDoodle - Thanks for the report! What type of RAM/CPU does your PicoShare server have? What is the server's CPU and RAM utilization while the download is happening?

**Comment by mtlynch** [comment]
>If needed, I can attempt to download again and get the stats from the container itself.

Yeah, that would be helpful. Is it possible the container did have limits somehow? That's a super beefy machine, so I don't know what the bottleneck would be to push it down to 500 KB/s. I'm assuming network wasn't saturated and disk wasn't being pegged at the time, either, right?

Is this with PicoShare 1.4.5?
